Use python debug adapter when in experiment

karthiknadig opened the pull request:
- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Appropriate comments and documentation strings in the code
- [x] Has sufficient logging.
- [x] Has telemetry for enhancements.
- [x] Unit tests & system/integration tests are added/updated

kimadeline commented on pythonFiles/install_ptvsd.py:
```diff
@@ -1,63 +1,57 @@
+EXTENSION_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+DEBUGGER_DEST = os.path.join(EXTENSION_ROOT, "pythonFiles", "lib", "python")
+DEBUGGER_PACKAGE = "ptvsd"
+DEBUGGER_VERSION = "5.0.0a5"
```
Do we need to do anything with the ptvsd version in `requirements.txt` (for example revert it to 4.3.2)?

karthiknadig replied:
Removing `ptvsd`, `packaging`, and `six` from `requirements.txt`

kimadeline commented on src/client/debugger/extension/adapter/factory.ts:
```diff
@@ -28,20 +23,29 @@ export class DebugAdapterDescriptorFactory implements IDebugAdapterDescriptorFac
+        if (this.experimentsManager.inExperiment(DebugAdapterNewPtvsd.experiment)) {
+            if (configuration.request === 'attach') {
+                const port = configuration.port ? configuration.port : 0;
+                if (port === 0) {
```
Is it possible to have a configuration with port === 0?

```suggestion
                if (port === undefined) {
```

and further down
```ts
return new DebugAdapterServer(configuration.port, configuration.host);
```

karthiknadig replied:
Users can put `port=0` in the launch json. It is not a valid value when connecting as a client.

kimadeline commented on src/client/debugger/extension/adapter/factory.ts:
```diff
@@ -73,15 +72,8 @@ export class DebugAdapterDescriptorFactory implements IDebugAdapterDescriptorFac
-            ptvsdPathToUse = path.join(EXTENSION_ROOT_DIR, 'pythonFiles');
-        }
-
+    public getPtvsdPath(): string {
```
You can remove this function, and call `path.join(EXTENSION_ROOT_DIR, 'pythonFiles', 'lib', 'python', 'ptvsd');` on L.43 directly.

karthiknadig replied:
I think this should use the getPtvsdPath API. Since that is used in other places like `api.ts`.

kimadeline replied:
doh thanks, i forgot that `getPtvsdPath()` was also called in `api.ts` 🤦‍♀ In that case you can combine the body of the function in one line, or if you want to be reallllll fancy you can do this:

```ts
public getPtvsdPath = () => path.join(EXTENSION_ROOT_DIR, 'pythonFiles', 'lib', 'python', 'ptvsd');
```

kimadeline commented on src/client/debugger/extension/adapter/factory.ts:
```diff
@@ -28,20 +23,29 @@ export class DebugAdapterDescriptorFactory implements IDebugAdapterDescriptorFac
+                    // If logToFile is set in the debug config then pass --log-dir <path-to-extension-dir> when launching the debug adapter.
+                    const logArgs = configuration.logToFile ? ['--log-dir', EXTENSION_ROOT_DIR] : [];
+                    const ptvsdPathToUse = this.getPtvsdPath();
+                    return new DebugAdapterExecutable(`${pythonPath}`, [path.join(ptvsdPathToUse, 'adapter'), ...logArgs]);
```
Wow what was I thinking there

```suggestion
                    return new DebugAdapterExecutable(pythonPath, [path.join(ptvsdPathToUse, 'adapter'), ...logArgs]);
```

kimadeline commented on src/test/debugger/extension/adapter/factory.unit.test.ts:
```diff
@@ -47,7 +44,6 @@ suite('Debugging - Adapter Factory', () => {
     let experimentsManager: ExperimentsManager;
     let spiedInstance: ExperimentsManager;
-    let executionFactory: IPythonExecutionFactory;
     let stateFactory: IPersistentStateFactory;
```
🚮  Since we don't need to store anything in persistent state anymore, you can remove the declaration, instantiation and all calls using:
- `stateFactory`
- `debugAdapterPersistentState`
- `extensions`

kimadeline commented on src/test/debugger/extension/adapter/factory.unit.test.ts:
```diff
@@ -158,7 +150,7 @@ suite('Debugging - Adapter Factory', () => {
     test('Return the value of configuration.pythonPath as the current python path if it exists and if we are in the experiment', async () => {
         const session = createSession({ pythonPath });
-        const debugExecutable = new DebugAdapterExecutable(pythonPath, [path.join(ptvsdPath, 'ptvsd', 'adapter')]);
+        const debugExecutable = new DebugAdapterExecutable(pythonPath, [path.join(ptvsdPath, 'lib', 'python', 'ptvsd', 'adapter')]);
```
You can update `ptvsdPath` on line 57 to be `path.join(EXTENSION_ROOT_DIR, 'pythonFiles', 'lib', 'python')` directly (up to you if you want to add `ptvsd` and `adapter` to the `path.join()` call)

kimadeline commented on src/test/mocks/vsc/extHostedTypes.ts:
```diff
@@ -1920,6 +1920,17 @@ export namespace vscMockExtHostedTypes {
+
+        constructor(port: number, host?: string) {
+            this.port = port;
+            // @ts-ignore
```
What do you need it here for? I don't think it's needed.

```suggestion
```

ericsnowcurrently commented on src/client/debugger/extension/adapter/factory.ts:
```diff
@@ -73,16 +72,8 @@ export class DebugAdapterDescriptorFactory implements IDebugAdapterDescriptorFac
-
-        return path.join(ptvsdPathToUse, 'ptvsd');
+    public getPtvsdPath(): string {
+        return path.join(EXTENSION_ROOT_DIR, 'pythonFiles', 'lib', 'python', 'ptvsd');
```
At this point, why no just make this a module-level constant and drop the method?

kimadeline commented on src/client/debugger/extension/adapter/factory.ts:
```diff
@@ -28,20 +23,29 @@ export class DebugAdapterDescriptorFactory implements IDebugAdapterDescriptorFac
-            const logArgs = configuration.logToFile ? ['--log-dir', EXTENSION_ROOT_DIR] : [];
-            const ptvsdPathToUse = await this.getPtvsdPath(pythonPath);
-            return new DebugAdapterExecutable(`${pythonPath}`, [path.join(ptvsdPathToUse, 'adapter'), ...logArgs]);
+        if (this.experimentsManager.inExperiment(DebugAdapterNewPtvsd.experiment)) {
```
I understand why Karthik did it this way though, since it's how it's done in the other experiments we have. We can then clean them all at once when tackling https://github.com/microsoft/vscode-python/issues/7869 instead of having partially clean telemetry.

karthiknadig replied:
I will defer this to #7869.

karthiknadig commented on pythonFiles/install_ptvsd.py:
```diff
@@ -1,63 +1,57 @@
-        raise Exception("ptvsd requirement not found.")
 
+def _get_debugger_wheel_urls():
+    json_uri = "https://pypi.org/pypi/{0}/json".format(DEBUGGER_PACKAGE)
```
Agreed will change to f-strings.

karthiknadig commented on pythonFiles/install_ptvsd.py:
```diff
@@ -1,63 +1,57 @@
+EXTENSION_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+DEBUGGER_DEST = os.path.join(EXTENSION_ROOT, "pythonFiles", "lib", "python")
+DEBUGGER_PACKAGE = "ptvsd"
+DEBUGGER_VERSION = "5.0.0a5"
```
I want to keep the package-name, version, and pyhton-ABI together, because this is a non-standard way of installing a python package. Eventually I want this file to be the only file that installs the debugger or references the debugger version.

The reasons to not have `requirements.txt`:
1. Extracting versions from `requirements.txt` means yet another file to parse.
2. We have dependency on yet another package which is unnecessary. `packaging` to parse each line in requirements.
3. Requirements file cannot be used to  install multiple versions of the same package to different target locations. We are using CLI to install old ptvsd. So the assumption that a developer will be able to find the version thing by using requirements.txt is already wrong:
https://github.com/microsoft/vscode-python/blob/c5414c5ea3e68db670fbbd1200e5769cf5e246fa/build/ci/templates/steps/build.yml#L23-L24

https://github.com/microsoft/vscode-python/blob/c5414c5ea3e68db670fbbd1200e5769cf5e246fa/build/ci/templates/test_phases.yml#L97-L98

4. The version that gets installed by using `pip` (in the .yml) gets overwritten by this installer. Gives a false sense of what is really installed.

karthiknadig commented on pythonFiles/install_ptvsd.py:
```diff
@@ -1,63 +1,57 @@
+        return list(
+            r["url"]
+            for r in json_response["releases"][DEBUGGER_VERSION]
+            if _contains(r["url"], DEBUGGER_PYTHON_VERSIONS)
```
Will add a comment to clarify.

karthiknadig commented on pythonFiles/install_ptvsd.py:
```diff
@@ -1,63 +1,57 @@
-import urllib.request
-import sys
+import os
+import urllib.request as url_lib
```
Agreed

karthiknadig commented on pythonFiles/tests/debug_adapter/test_install_ptvsd.py:
```diff
@@ -0,0 +1,35 @@
+    reason="PTVSD wheels shipped for Python 3.7 only",
+)
+def test_install_ptvsd(tmpdir):
+    import install_ptvsd
```
This is to prevent `pytest` from trying to load `install_ptvsd` in python 2.7. In 2.7 `import urllib.request` will fail.

karthiknadig commented on src/client/debugger/extension/adapter/factory.ts:
```diff
@@ -28,20 +23,29 @@ export class DebugAdapterDescriptorFactory implements IDebugAdapterDescriptorFac
-            const ptvsdPathToUse = await this.getPtvsdPath(pythonPath);
-            return new DebugAdapterExecutable(`${pythonPath}`, [path.join(ptvsdPathToUse, 'adapter'), ...logArgs]);
+        if (this.experimentsManager.inExperiment(DebugAdapterNewPtvsd.experiment)) {
+            if (configuration.request === 'attach') {
```
The current design is broken. Attach request will fail, if it is done using same way as `launch`.

karthiknadig commented:
I plan on merging this after the CI is clean. Some bad merge in master has the build hygiene failing everywhere.
1. I have addressed all the comments. except for the one about `requirements.txt` and I have provided the reason why we should **not** use it for ptvsd.
2. This is needed to make progress on the reload work.
